kafka/producer-manager.go

```diff
@@ -0,0 +1,144 @@
+package kafka
+
+import (
+	"sync"
+
+	"time"
+
+	"net/http"
+
+	"github.com/Shopify/sarama"
+	"github.com/Sirupsen/logrus"
+	"github.com/pressly/chi"
+	"github.com/thrawn01/args"
+	"golang.org/x/net/context"
+)
+
+type contextKey int
+
+const (
+	producerManagerKey contextKey = 1
+)
+
+func SetProducerManager(ctx context.Context, manager *ProducerManager) context.Context {
+	return context.WithValue(ctx, producerManagerKey, manager)
+}
+
+func GetProducerManager(ctx context.Context) *ProducerManager {
+	obj, ok := ctx.Value(producerManagerKey).(*ProducerManager)
+	if !ok {
+		panic("No kafka.ProducerManager found in context")
+	}
+	return obj
+}
+
+type ProducerManager struct {
+	done      chan struct{}
+	parser    *args.ArgParser
+	current   chan Producer
```

Comment:
As far as I know, this is not an established pattern, I've been thinking about this pattern for a while and this project seemed like a good place to experiment. I want to continue refining this pattern if it turns out I like it enough.  There is definitely a need to encapsulate some common logic, and I have some ideas on how to do this, if you are ok with it, I will make a new PR with my changes so as to not destroy these comments before everyone else has a chance to review the code. I actually prefer your suggestion of remembering the new connection in ```self.session```. It also turns out to be more preformant.  [Here is a benchmark test](https://github.com/thrawn01/args/blob/master/benchmark/benchmark.go) I wrote to judge the merits of using channels for single value access over mutexes. I hope to write a blog post publishing my findings.